### DISCUSSION TITLE
Collections grouped maps

### DISCUSSION THREAD

**Discussion by suhtai** [arrow-kt/arrow#2390] [Ideas]
I'm not sure if this is a known use case on other Functional libs but I was wondering if we could have a way to map (and `flatMap`) a collection to another type. I'm not sure what the name could be as it could be considered a comprehension for limiting the size of a list and doing a fold on that.

Something like this could work for Collections:

```kotlin
public fun <T, R> Collection<T>.takeMap(transform: (T, T) -> R): R? =
    take(2).takeIf { it.size == 2 }?.let { (a, b) -> transform(a, b) }

public fun <T, R> Collection<T>.takeMap(transform: (T) -> R): R? =
    take(1).takeIf { it.size == 1 }?.let { (a) -> transform(a) }
```

and then we can use it like:

```kotlin
val message: String? = list.takeMap { first, second -> "$first and $second" }
```

We can handle any arity thanks to overriding so the caller can decide how many items to look for by the number of parameters on the closure.

For Lists we could also do these for the last items:

```kotlin
public fun <T, R> List<T>.mapLast(transform: (T, T) -> R): R? =
    takeLast(2).takeIf { it.size == 2 }?.let { (a, b) -> transform(a, b) }

public fun <T, R> List<T>.mapLast(transform: (T) -> R): R? =
    takeLast(1).takeIf { it.size == 1 }?.let { (a) -> transform(a) }
```

And these could also made to fit flatMap.

Our use case for these is as follows:

We have a list of subscriptions for a user and want to check that we have at least one that is active, so we do something like:

```kotlin
val User.activeSubscription: SubscriptionId? 
  get() = subscriptions.mapLast { subscription -> subscription.id }
```

**Comment by suhtai** [comment]
Further discussions from [Slack](https://kotlinlang.slack.com/archives/C8UK6RTHU/p1620127970049100)

**Comment by kyay10** [comment]
I'd worry about calling this `map`, since `map` means doing something for all items.
